pkg/reconciler/ingress/ingress.go

```diff
@@ -0,0 +1,310 @@
+		// a cluster.
+		//
+		// This update can come from the creation or because the syncer has update the status.
+
+		rootIngressName := ingress.Labels[ownedByLabel]
+		// A leaf Ingress was updated; get others and aggregate status.
+		ownedBySelector := fmt.Sprintf("%s=%s", ownedByLabel, rootIngressName)
+		sel, err := labels.Parse(ownedBySelector)
+		if err != nil {
+			return err
+		}
+		others, err := c.ingressLister.List(sel)
+		if err != nil {
+			return err
+		}
+
+		// Get the rootIngress based on the labels.
+		var rootIngress *networkingv1.Ingress
+
+		rootIf, exists, err := c.ingressIndexer.Get(&networkingv1.Ingress{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace:   ingress.Namespace,
+				Name:        rootIngressName,
+				ClusterName: ingress.GetClusterName(),
+			},
+		})
+		if err != nil {
+			return err
+		}
+
+		// TODO(jmprusi): A leaf without rootIngress?
+		if !exists {
+			//TODO(jmprusi): Add user-facing condition to leaf.
+			klog.Errorf("root Ingress not found %s/%s/%s", ingress.ClusterName, ingress.Namespace, rootIngressName)
+			return nil
+		}
+
+		rootIngress = rootIf.(*networkingv1.Ingress).DeepCopy()
+
+		// Clean the current status, and then recreate it from the other leafs.
+		rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{}
+		for _, o := range others {
+			rootIngress.Status.LoadBalancer.Ingress = append(rootIngress.Status.LoadBalancer.Ingress, o.Status.LoadBalancer.Ingress...)
+		}
+
+		// If the envoy controlplane is enabled, we update the cache and generate and send to envoy a new snapshot.
+		if c.envoycontrolplane != nil {
+			statusHost := generateStatusHost(c.domain, rootIngress)
+			// Now overwrite the Status of the rootIngress with our desired LB
+			rootIngress.Status.LoadBalancer.Ingress = []corev1.LoadBalancerIngress{{
+				Hostname: statusHost,
+			}}
+
+			// We want the controlplane to pick this one up and update the envoy config.
+			ingress.Labels[toEnvoyLabel] = "true"
+		}
+
+		// Update the rootIngress status with our desired LB.
+		//TODO(jmprusi): Use patch instead of update.
+		if _, err := c.client.Cluster(rootIngress.ClusterName).NetworkingV1().Ingresses(rootIngress.Namespace).UpdateStatus(ctx, rootIngress, metav1.UpdateOptions{}); err != nil {
+			klog.Errorf("failed to update root ingress status: %v", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *Controller) updateLeafs(ctx context.Context, currentLeaves []*networkingv1.Ingress, desiredLeaves []*networkingv1.Ingress) ([]*networkingv1.Ingress, error) {
+	// Let's iterate over the current leaves and remove the ones that are not in the desired leaves, and update the ones that match.
+	for _, currentleaf := range currentLeaves {
+		found := false
+
+		// We will remove a leaf from the desiredLeaves if it's found already, so we don't create it later.
+		for i, rlen := 0, len(desiredLeaves); i < rlen; i++ {
+			j := i - (rlen - len(desiredLeaves))
```

Comment:
This is confusing. I also think that given this function is currently responsible for updating existing leaves and deleting nondesired ones, it would make sense to move the creation of "missing" leaves in here as well. If we want to split it up later, we can, but I think the current division of "update+delete, return missing" and "create missing" is not very clear.